pkg/web/oauthclient/callback.go

```diff
@@ -134,7 +134,10 @@ func (oc *OAuthClient) HandleCallback(w http.ResponseWriter, r *http.Request) {
 	}
 
 	if err := oc.callback(w, r, token, stateCookie.Next); err != nil {
-		webhandlers.Error(w, r, errExchange.WithCause(err))
+		if _, ok := errors.From(err); !ok {
+			err = errExchange.WithCause(err)
+		}
```

Comment:
The default callback never returns an error. So this errors only when there's a custom callback configured. In that case, return the error without wrapping it in an "token exchange refused" permission denied error. The callback is not about the token exchange, and the callback may fail because of other reasons than permission denied.